File: csharp/Tdcv2/Parser/PairedData.cs
using System.Text;

namespace Tdcv2.Parser;

/// <summary>
/// Paired raw text, rewritten before the lexer ever sees it.
/// </summary>
/// <remarks>
/// <para>
/// <c>&lt;data pair="X"&gt;…&lt;/data pair="X"&gt;</c> lets a body carry a literal
/// <c>&lt;/data&gt;</c> — a snippet of TDC syntax inside generated documentation, say. The grammar
/// keeps one static <c>&lt;/data&gt;</c> close token because a lexer that had to know which closer
/// belongs to which opener would need the pair value inside a token rule, so the pairing is
/// resolved here instead: the paired closer becomes a plain <c>&lt;/data&gt;</c> and every literal
/// <c>&lt;/data&gt;</c> in the body becomes a sentinel the lexer reads as ordinary text.
/// <see cref="Restore"/> puts the sentinel back when a body is read.
/// </para>
/// <para>
/// The rewrite is length-preserving on purpose. Everything the lexer, the parser and the validator
/// report afterwards carries a line and a column, and those have to point into the file the user
/// wrote rather than into the one this pass produced — which is why the closing tag's leftover
/// characters become spaces instead of disappearing.
/// </para>
/// <para>
/// Ported from <c>typescript/src/parser/paired-data.ts</c>. The five implementations have to agree
/// character for character, malformed input included, so this follows the reference's decisions
/// even where a fresh design would choose otherwise.
/// </para>
/// </remarks>
public static class PairedData
{
    /// <summary>
    /// NUL, which cannot appear in a hand-written config, standing in for the two ends of a
    /// literal <c>&lt;/data&gt;</c> while the lexer runs.
    /// </summary>
    private const char Guard = '\0';

    /// <summary>
    /// What a literal <c>&lt;/data&gt;</c> inside a paired body becomes for the duration of
    /// lexing. Exactly as long as the text it stands in for, which is what keeps every later
    /// position honest.
    /// </summary>
    private static readonly string Sentinel = Guard + "/data" + Guard;

    private const string Open = "<data";
    private const string Close = "</data>";
    private const string ClosePrefix = "</data";

    /// <summary>One paired tag that does not line up, at the position a person can act on.</summary>
    public sealed record Problem(int Line, int Column, string Message);

    /// <summary>The source to lex, and everything wrong with the paired tags in it.</summary>
    public sealed record Rewrite(string Source, IReadOnlyList<Problem> Problems);

    /// <summary>A close tag, and the <c>pair</c> it carried if it carried one.</summary>
    private readonly record struct CloseTag(int Start, int End, string? Pair);

    /// <summary>A line (1-based) and column (0-based), as every diagnostic here counts them.</summary>
    private readonly record struct Position(int Line, int Column);

    public static Rewrite Preprocess(string source)
    {
        var output = new StringBuilder();
        int cursor = 0;
        var problems = new List<Problem>();
        var seen = new Dictionary<string, Position>(StringComparer.Ordinal);

        while (cursor < source.Length)
        {
            int openStart = source.IndexOf(Open, cursor, StringComparison.Ordinal);
            if (openStart < 0)
            {
                output.Append(source, cursor, source.Length - cursor);
                break;
            }

            if (!IsDataOpenAt(source, openStart))
            {
                // `<database>` and friends: emit the false start and keep looking past it.
                output.Append(source, cursor, openStart + Open.Length - cursor);
                cursor = openStart + Open.Length;
                continue;
            }

            int openEnd = FindTagEnd(source, openStart);
            if (openEnd < 0)
            {
                output.Append(source, cursor, source.Length - cursor);
                break;
            }

            string openText = source[openStart..(openEnd + 1)];
            string? pair = PairValue(openText);
            if (pair is null || IsSelfClosing(openText))
            {
                output.Append(source, cursor, openEnd + 1 - cursor);
                cursor = openEnd + 1;
                continue;
            }

            Position pairPosition = At(source, openStart + openText.IndexOf(pair, StringComparison.Ordinal));
            if (seen.TryGetValue(pair, out Position previous))
            {
                problems.Add(new Problem(
                    pairPosition.Line,
                    pairPosition.Column,
                    $"duplicate <data pair=\"{pair}\"> value. "
                    + $"First use was at line {previous.Line}, column {previous.Column}."));
            }
            else
            {
                seen[pair] = pairPosition;
            }

            int bodyStart = openEnd + 1;
            (CloseTag? match, CloseTag? mismatch) = FindClose(source, bodyStart, pair);
            if (match is null)
            {
                Position at = At(source, mismatch?.Start ?? openStart);
                string message = mismatch is null
                    ? $"unclosed <data pair=\"{pair}\">"
                    : $"expected </data pair=\"{pair}\">, got </data pair=\"{mismatch.Value.Pair}\">";
                problems.Add(new Problem(at.Line, at.Column, message));

                // Nothing after an unmatched opener can be rewritten with any confidence about
                // where the body was meant to end, so the rest of the file is handed over untouched.
                output.Append(source, cursor, source.Length - cursor);
                break;
            }

            CloseTag close = match.Value;
            output.Append(source, cursor, bodyStart - cursor);
            output.Append(source[bodyStart..close.Start].Replace(Close, Sentinel, StringComparison.Ordinal));
            output.Append(Close);
            output.Append(StructuralWhitespace(source[(close.Start + Close.Length)..(close.End + 1)]));
            cursor = close.End + 1;
        }

        return new Rewrite(output.ToString(), problems);
    }

    /// <summary>A body as the user wrote it, with the sentinel back to a literal close tag.</summary>
    public static string Restore(string text) =>
        text.Replace(Sentinel, Close, StringComparison.Ordinal);

    private static bool IsDataOpenAt(string source, int index)
    {
        int after = index + Open.Length;
        if (after >= source.Length)
        {
            return true;
        }
        char next = source[after];
        return next == '>' || next == '/' || IsSpace(next);
    }

    private static bool IsSelfClosing(string tagText)
    {
        int at = tagText.Length - 2; // The tag always ends in '>'; read back from there.
        while (at >= 0 && IsSpace(tagText[at]))
        {
            at--;
        }
        return at >= 0 && tagText[at] == '/';
    }

    /// <summary>
    /// The close that pairs with <paramref name="expected"/>, or — when there is none — the first
    /// close that carried some other pair, which is the difference between "you closed it wrong"
    /// and "you never closed it".
    /// </summary>
    private static (CloseTag? Match, CloseTag? Mismatch) FindClose(
        string source, int start, string expected)
    {
        int searchAt = start;
        CloseTag? mismatch = null;

        while (searchAt < source.Length)
        {
            int closeStart = source.IndexOf(ClosePrefix, searchAt, StringComparison.Ordinal);
            if (closeStart < 0)
            {
                break;
            }
            int closeEnd = FindTagEnd(source, closeStart);
            if (closeEnd < 0)
            {
                break;
            }

            string? closePair = PairValue(source[closeStart..(closeEnd + 1)]);
            if (string.Equals(closePair, expected, StringComparison.Ordinal))
            {
                return (new CloseTag(closeStart, closeEnd, closePair), mismatch);
            }
            if (closePair is not null && mismatch is null)
            {
                mismatch = new CloseTag(closeStart, closeEnd, closePair);
            }
            searchAt = closeStart + ClosePrefix.Length;
        }

        return (null, mismatch);
    }

    /// <summary>The '>' that ends a tag, ignoring any inside quotes so <c>if="a&gt;b"</c> does not.</summary>
    private static int FindTagEnd(string source, int start)
    {
        char quote = '\0';
        for (int at = start; at < source.Length; at++)
        {
            char ch = source[at];
            if (quote != '\0')
            {
                if (ch == quote)
                {
                    quote = '\0';
                }
                continue;
            }
            if (ch == '"' || ch == '\'')
            {
                quote = ch;
                continue;
            }
            if (ch == '>')
            {
                return at;
            }
        }
        return -1;
    }

    /// <summary>
    /// The <c>pair="…"</c> value in a tag, as the reference's <c>\bpair\s*=\s*"([^"\r\n]*)"</c>.
    /// </summary>
    private static string? PairValue(string tagText)
    {
        int at = 0;
        while (true)
        {
            int found = tagText.IndexOf("pair", at, StringComparison.Ordinal);
            if (found < 0)
            {
                return null;
            }

            // The word boundary: `superpair=` is not a pair attribute, `data-pair=` is.
            if (found > 0 && IsWordChar(tagText[found - 1]))
            {
                at = found + 1;
                continue;
            }

            int scan = SkipSpace(tagText, found + "pair".Length);
            if (scan >= tagText.Length || tagText[scan] != '=')
            {
                at = found + 1;
                continue;
            }
            scan = SkipSpace(tagText, scan + 1);
            if (scan >= tagText.Length || tagText[scan] != '"')
            {
                at = found + 1;
                continue;
            }

            scan++;
            int valueStart = scan;
            while (scan < tagText.Length
                && tagText[scan] != '"' && tagText[scan] != '\r' && tagText[scan] != '\n')
            {
                scan++;
            }
            if (scan < tagText.Length && tagText[scan] == '"')
            {
                return tagText[valueStart..scan];
            }
            at = found + 1;
        }
    }

    private static int SkipSpace(string text, int at)
    {
        while (at < text.Length && IsSpace(text[at]))
        {
            at++;
        }
        return at;
    }

    /// <summary>Line breaks kept, everything else blanked — the leftovers hold their place.</summary>
    private static string StructuralWhitespace(string text)
    {
        var output = new StringBuilder();
        for (int at = 0; at < text.Length; at++)
        {
            char ch = text[at];
            if (ch == '\n' || ch == '\r')
            {
                output.Append(ch);
                continue;
            }
            output.Append(' ');

            // By code point, not by char: an astral character in a pair value is one space in the
            // reference, and two would push every column after it out by one.
            if (char.IsHighSurrogate(ch) && at + 1 < text.Length && char.IsLowSurrogate(text[at + 1]))
            {
                at++;
            }
        }
        return output.ToString();
    }

    private static Position At(string source, int index)
    {
        int line = 1;
        int column = 0;
        for (int at = 0; at < index; at++)
        {
            if (source[at] == '\n')
            {
                line++;
                column = 0;
            }
            else
            {
                column++;
            }
        }
        return new Position(line, column);
    }

    /// <summary>
    /// Whitespace as JavaScript's <c>\s</c> defines it, which is what the reference tests against.
    /// Spelling the set out is what stops five languages disagreeing over an exotic space:
    /// <c>char.IsWhiteSpace</c> admits U+0085 and refuses U+FEFF, so it is not this set under
    /// another name.
    /// </summary>
    private static bool IsSpace(char ch) =>
        ch is '\t' or '\n' or '\v' or '\f' or '\r' or ' '
            or '\u00a0' or '\u1680' or '\u2028' or '\u2029'
            or '\u202f' or '\u205f' or '\u3000' or '\ufeff'
        || (ch >= '\u2000' && ch <= '\u200a');

    private static bool IsWordChar(char ch) =>
        (ch >= 'a' && ch <= 'z')
        || (ch >= 'A' && ch <= 'Z')
        || (ch >= '0' && ch <= '9')
        || ch == '_';
}

File: python/src/tdcv2/parser/paired_data.py
"""Paired raw text, rewritten before the lexer ever sees it.

``<data pair="X">…</data pair="X">`` lets a body carry a literal ``</data>`` — a snippet of TDC
syntax inside generated documentation, say. The grammar keeps one static ``</data>`` close token
because a lexer that had to know which closer belongs to which opener would need the pair value
inside a token rule, so the pairing is resolved here instead: the paired closer becomes a plain
``</data>`` and every literal ``</data>`` in the body becomes a sentinel the lexer reads as
ordinary text. :func:`restore` puts the sentinel back when a body is read.

The rewrite is length-preserving on purpose. Everything the lexer, the parser and the validator
report afterwards carries a line and a column, and those have to point into the file the user
wrote rather than into the one this pass produced — which is why the closing tag's leftover
characters become spaces instead of disappearing.

Ported from ``typescript/src/parser/paired-data.ts``. The five implementations have to agree
character for character, malformed input included, so this follows the reference's decisions even
where a fresh design would choose otherwise.
"""

from __future__ import annotations

#: What a literal ``</data>`` inside a paired body becomes for the duration of lexing. Exactly as
#: long as the text it stands in for, which is what keeps every later position honest.
SENTINEL = "\u0000/data\u0000"

_OPEN = "<data"
_CLOSE = "</data>"
_CLOSE_PREFIX = "</data"

#: The reference tests whitespace with JavaScript's ``\s``. Spelling the set out is what stops
#: five languages disagreeing over an exotic space character, since none of their built-in
#: whitespace predicates covers exactly this set.
_SPACE = frozenset("\t\n\v\f\r \u00a0\u1680\u2028\u2029\u202f\u205f\u3000\ufeff") | {
    chr(code) for code in range(0x2000, 0x200B)
}

_WORD = frozenset("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789_")


def preprocess(source: str) -> tuple[str, list[tuple[int, int, str]]]:
    """The source to lex, plus ``(line, column, message)`` for every paired tag that is wrong."""
    out: list[str] = []
    cursor = 0
    problems: list[tuple[int, int, str]] = []
    seen: dict[str, tuple[int, int]] = {}

    while cursor < len(source):
        open_start = source.find(_OPEN, cursor)
        if open_start < 0:
            out.append(source[cursor:])
            break

        if not _is_data_open_at(source, open_start):
            # `<database>` and friends: emit the false start and keep looking past it.
            out.append(source[cursor : open_start + len(_OPEN)])
            cursor = open_start + len(_OPEN)
            continue

        open_end = _find_tag_end(source, open_start)
        if open_end < 0:
            out.append(source[cursor:])
            break

        open_text = source[open_start : open_end + 1]
        pair = _pair_value(open_text)
        if pair is None or _is_self_closing(open_text):
            out.append(source[cursor : open_end + 1])
            cursor = open_end + 1
            continue

        pair_position = _position(source, open_start + open_text.index(pair))
        previous = seen.get(pair)
        if previous is not None:
            problems.append(
                (
                    pair_position[0],
                    pair_position[1],
                    f'duplicate <data pair="{pair}"> value. '
                    f"First use was at line {previous[0]}, column {previous[1]}.",
                )
            )
        else:
            seen[pair] = pair_position

        body_start = open_end + 1
        match, mismatch = _matching_close(source, body_start, pair)
        if match is None:
            if mismatch is not None:
                line, column = _position(source, mismatch[0])
                message = f'expected </data pair="{pair}">, got </data pair="{mismatch[2]}">'
            else:
                line, column = _position(source, open_start)
                message = f'unclosed <data pair="{pair}">'
            problems.append((line, column, message))
            # Nothing after an unmatched opener can be rewritten with any confidence about where
            # the body was meant to end, so the rest of the file is handed over untouched.
            out.append(source[cursor:])
            break

        close_start, close_end = match
        out.append(source[cursor:body_start])
        out.append(source[body_start:close_start].replace(_CLOSE, SENTINEL))
        out.append(_CLOSE)
        out.append(_structural_whitespace(source[close_start + len(_CLOSE) : close_end + 1]))
        cursor = close_end + 1

    return "".join(out), problems


def restore(text: str) -> str:
    """A ``<data>`` body as the user wrote it, with the sentinel back to a literal ``</data>``."""
    return text.replace(SENTINEL, _CLOSE)


def _is_data_open_at(source: str, index: int) -> bool:
    after = index + len(_OPEN)
    if after >= len(source):
        return True
    return source[after] in ">/" or source[after] in _SPACE


def _is_self_closing(tag_text: str) -> bool:
    at = len(tag_text) - 2  # The tag always ends in '>'; look at what leads up to it.
    while at >= 0 and tag_text[at] in _SPACE:
        at -= 1
    return at >= 0 and tag_text[at] == "/"


def _matching_close(
    source: str, start: int, expected: str
) -> tuple[tuple[int, int] | None, tuple[int, int, str] | None]:
    """The close that pairs with ``expected``, else the first close that carries a different one."""
    search_at = start
    mismatch: tuple[int, int, str] | None = None

    while search_at < len(source):
        close_start = source.find(_CLOSE_PREFIX, search_at)
        if close_start < 0:
            break
        close_end = _find_tag_end(source, close_start)
        if close_end < 0:
            break

        close_pair = _pair_value(source[close_start : close_end + 1])
        if close_pair == expected:
            return (close_start, close_end), mismatch
        if close_pair is not None and mismatch is None:
            mismatch = (close_start, close_end, close_pair)
        search_at = close_start + len(_CLOSE_PREFIX)

    return None, mismatch


def _find_tag_end(source: str, start: int) -> int:
    """The '>' that ends a tag, ignoring any inside quotes so ``if="a>b"`` does not end it early."""
    quote: str | None = None
    for at in range(start, len(source)):
        char = source[at]
        if quote is not None:
            if char == quote:
                quote = None
            continue
        if char in "\"'":
            quote = char
            continue
        if char == ">":
            return at
    return -1


def _pair_value(tag_text: str) -> str | None:
    """The ``pair="…"`` value in a tag, as the reference's ``\\bpair\\s*=\\s*"([^"\\r\\n]*)"``."""
    length = len(tag_text)
    at = 0
    while True:
        found = tag_text.find("pair", at)
        if found < 0:
            return None
        # The word boundary: `superpair=` is not a pair attribute, `data-pair=` is.
        if found > 0 and tag_text[found - 1] in _WORD:
            at = found + 1
            continue

        scan = _skip_space(tag_text, found + 4)
        if scan >= length or tag_text[scan] != "=":
            at = found + 1
            continue
        scan = _skip_space(tag_text, scan + 1)
        if scan >= length or tag_text[scan] != '"':
            at = found + 1
            continue

        scan += 1
        value_start = scan
        while scan < length and tag_text[scan] not in '"\r\n':
            scan += 1
        if scan < length and tag_text[scan] == '"':
            return tag_text[value_start:scan]
        at = found + 1


def _skip_space(text: str, at: int) -> int:
    while at < len(text) and text[at] in _SPACE:
        at += 1
    return at


def _structural_whitespace(text: str) -> str:
    """Line breaks kept, everything else blanked — the closer's leftovers hold their place."""
    return "".join(char if char in "\n\r" else " " for char in text)


def _position(source: str, index: int) -> tuple[int, int]:
    line = 1
    column = 0
    for at in range(index):
        if source[at] == "\n":
            line += 1
            column = 0
        else:
            column += 1
    return line, column
